main/lsp/requests/workspace_symbols.cc

```diff
@@ -21,12 +44,134 @@ unique_ptr<SymbolInformation> symbolRef2SymbolInformation(const LSPConfiguration
+            score += symbolCharsConsumed;
+        }
+        if (symbolIter == symbolEnd) {
+            break;
+        }
+    }
+    return make_pair(score, queryMatchLength);
+}
+} // namespace
+
+vector<unique_ptr<SymbolInformation>> SymbolMatcher::doQuery(string_view query_view, int limit) {
+    vector<unique_ptr<SymbolInformation>> results;
+    string query(query_view);
+    if (query.empty()) {
+        return results;
+    }
+    const int queryLength = query.length();
+
+    struct ScoreInfo {
+        u4 symbolIndex = 0;
+        int lengthMatched = 0;
+        int score = 0;
+    };
+    vector<ScoreInfo> scoreInfos(gs.symbolsUsed());
+    // First pass: match short names
+    for (u4 symbolIndex = 1; symbolIndex < gs.symbolsUsed(); symbolIndex++) {
+        auto symbolData = core::SymbolRef(gs, symbolIndex).data(gs);
+        auto nameData = symbolData->name.data(gs);
+        if (nameData->kind == core::NameKind::UNIQUE) {
+            continue;
+        }
+        auto shortName = nameData->shortName(gs);
+        auto [score, lengthMatched] = scoreSymbol(shortName, query);
+        scoreInfos[symbolIndex].symbolIndex = symbolIndex;
+        scoreInfos[symbolIndex].score = score;
+        scoreInfos[symbolIndex].lengthMatched = lengthMatched;
+    }
+
+    // Second pass: in some cases, try again if matching long name might improve the result
+    for (u4 symbolIndex = 1; symbolIndex < gs.symbolsUsed(); symbolIndex++) {
+        ScoreInfo &scoreInfo = scoreInfos[symbolIndex];
+        if (scoreInfo.symbolIndex == 0) {
+            continue; // symbol ineligible
+        }
+        auto symbolData = core::SymbolRef(gs, symbolIndex).data(gs);
+        auto ownerRef = symbolData->owner;
+        if (!ownerRef.exists()) {
+            continue; // no owner
+        }
+        ScoreInfo &ownerScoreInfo = scoreInfos[ownerRef._id];
+        if (ownerScoreInfo.lengthMatched == 0) {
+            continue; // owner can't improve search result
+        }
+        if (ownerScoreInfo.lengthMatched == queryLength) {
+            continue; // don't bother retrying: it fully matches the container
+        }
+        if (ownerScoreInfo.lengthMatched <= scoreInfo.lengthMatched && ownerScoreInfo.score >= scoreInfo.score) {
+            continue; // owner's score was same-or-worse on equal-or-shorter match, don't bother
+        }
+        auto fullName = symbolData->showFullName(gs);
+        auto [score, lengthMatched] = scoreSymbol(fullName, query);
+        if (lengthMatched > scoreInfo.lengthMatched ||
+            (lengthMatched == scoreInfo.lengthMatched && score <= scoreInfo.score)) {
+            // Found a better match (longer or better score)
+            scoreInfo.score = score;
+            scoreInfo.lengthMatched = lengthMatched;
+        }
+    }
+    fast_sort(scoreInfos, [queryLength](ScoreInfo &left, ScoreInfo &right) -> bool {
+        int leftScore = (left.lengthMatched != queryLength) ? INT_MAX : left.score;
```

Comment:
Is there a reason why this check is pushed down here? (Can we easily assign INT_MAX to `left.score` in the first place?) That would clean up this code + the loop below.

Reply:
Am thinking that this was a bad idea in general because it makes `fast_sort` operate on O(all-symbols) rather than O(matching-symbols).  I did it for a misguided reason that I was trying to minimize storage, I think the next version will be much cleaner.